### PR TITLE
Establish that read holds are valid as acquired

### PR DESCRIPTION
**Pull request by frankmcsherry**
Light diagnostic assert that should tell us if we initialize read holds that are not valid at the moment we establish them.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/adapter/src/coord/read_policy.rs [hunk 1]
```diff
@@ -145,8 +145,18 @@ impl<S: Append + 'static> crate::coord::Coordinator<S> {
...
+                        .compute(*compute_instance)
+                        .unwrap()
+                        .collection(*id)
+                        .unwrap()
+                        .read_capabilities
+                        .frontier()
+                        .to_owned();
                     let TimelineState { read_holds, .. } =
                         self.ensure_timeline_state(timeline).await;
+                    assert!(initial_frontier.less_equal(&read_holds.time));
```
Comment (jkosh44) [inline src/adapter/src/coord/read_policy.rs:159]: We probably want a more detailed error message like in the previous PR.

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/adapter/src/coord/read_policy.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/adapter/src/coord/read_policy.rs:159]: Done! Now we'll be able to see in gory detail why the invariant does not hold!

**Comment by frankmcsherry** [comment]
This is crashing all over the place, with errors like (fast slt):
```
thread 'coordinator' panicked at 'Storage collection User(1) has read frontier Antichain { elements: [1659398239926] } not less-equal to read_hold.time: 1659398239864', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-061e6fee039b16ee4-1/materialize/tests/src/adapter/src/coord/read_policy.rs:207:17
```
Per @jkosh44 suggestion, this error goes away if we remove some table initialization code that marks tables as invalid through some initial `since_ts`:
```
diff --git a/src/adapter/src/coord/sequencer.rs b/src/adapter/src/coord/sequencer.rs
index 1e68e0d3b..2a1958199 100644
--- a/src/adapter/src/coord/sequencer.rs
+++ b/src/adapter/src/coord/sequencer.rs
@@ -864,13 +864,6 @@ impl<S: Append + 'static> Coordinator<S> {
                     .await
                     .unwrap();
 
-                let policy = ReadPolicy::ValidFrom(Antichain::from_elem(since_ts));
-                self.controller
-                    .storage_mut()
-                    .set_read_policy(vec![(table_id, policy)])
-                    .await
-                    .unwrap();
-
                 self.initialize_storage_read_policies(
                     vec![table_id],
                     DEFAULT_LOGICAL_COMPACTION_WINDOW_MS,
```
I'm not sure this is the fix we want, though: instead, we should probably advance the time of the read hold, if this is permitted (e.g. not in a transaction). Otherwise, we are in a bit of a pickle with the representation of a single time for groups of read holds.

**Comment by frankmcsherry** [comment]
I've changed the `assert!`s to `tracing::error!`s. This should allow this to land, and us to start to figure out what is going on with folks who experience us (we haven't been able to reproduce ourselves).

**Pull request merged**